src/traits.rs

```diff
@@ -99,20 +110,84 @@ impl Lerp for u32 {
     type Ratio = f64;
 }
 
+impl Lerp for f32 {
+    type Ratio = f32;
+
+    fn lerp(a: Self, b: Self, ratio: Self::Ratio) -> Self {
+        a + (b - a) * ratio
+    }
+}
+
+
+pub trait Sample: Primitive + 'static {
+    const MAX_SAMPLE_VALUE: Self;
+    const RGB_COLOR_TYPE: ColorTypeOrErr;
+    const RGBA_COLOR_TYPE: ColorTypeOrErr;
+    const L_COLOR_TYPE: ColorTypeOrErr;
+    const LA_COLOR_TYPE: ColorTypeOrErr;
```

Comment:
Usage of an associated constant brings up an interesting issue that wasn't obvious to me from the start, related to forward compatibility. We can't really change these constants in a minor version without risk of breakage. At least we don't have a guarantee here. This is because downstream users might have code such as this: 

```rust
<u32 as Sample>::L_COLOR_TYPE.is_err()
```

[Playground for trying it out](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4d812535bcbac5b0446022ae375db23c)

That's quite unfortunate, isn't it? The only real remedy would be _hiding_ the existance of these constants by putting them in a public-private trait.

But of course the same would apply to the constants in `Luma<_> as Pixel`. Of course there we could change them to not be constants afterall, but just have static accessor methods. But what's the cost associated with that?

Reply:
I think the solution with the associated constant has a heap of problems. It's not a clean design, it's a rigid architecture that makes future changes harder. The clean solution would be to change the `Pixel` trait.

The example you linked in the playground shows one of the symptoms. However, if we cannot change the pixel trait, this is an acceptable trade-off. If we need to change it, let it be a breaking change. That's the sacrifice we need to make by not cleaning up the pixel trait

Reply:
Pretty much exactly my throughts. However, we are in the fortunate situation of being able to break the pixel trait. We should only be careful and keep the changes as small as possible.

Reply:
The last time I tried to come up with a clean solution, I ended up completely reworking the pixel trait. I don't know whether I can come up with a smaller refactor, but I can try

Reply:
What if we move the associated const `COLOR_TYPE` into a subtrait, something like `ColoredPixel`.

Reply:
Because I'm decently sure the constant isn't actually being used in the _type system_ (how could it?) so it should be generally feasible to add that extra bound to usage sites.

Reply:
The `COLOR_TYPE` constant is used in the `impl trait Pixel for $T` macro. It's necessary because the Pixel trait defines the constant `COLOR_TYPE: ColorTypeOrErr`.

I'll try separating them

Reply:
Yes, I know that's what Pixel is doing but it's not a functionality. That's just the same question again on a different trait. But to what effect?